File: controllers/admin/posts.rb
class Blog
  class Admin
    class Posts < Harbor::Controller
      
      get "new" do
        @post = ::Post.new
        render "admin/posts/edit"
      end
      
      get ":slug/edit" do |slug|
        @post = ::Post.db.get_by_slug slug
        render "admin/posts/edit"
      end
      
      post do |title, published_at, body, categories, _id = nil|
        @post = _id.blank? ? Post.new : Post.db.get_by_id(_id)
        @post.title = title
        @post.published_at = published_at
        @post.body = body
        @post.categories = categories.split(/,?\s+/)
        
        if @post.errors.empty?
          @post.save
          response.redirect "/admin"
        else
          render "admin/posts/edit"
        end
      end
      
      get ":id/delete" do |id|
        Post.db.delete(id)
        response.redirect "/admin"
      end
      
    end
  end
end

File: app/controllers/admin/Posts.scala
package controllers.admin

import play.api.mvc._
import models._
import controllers.Secured

object Posts extends Secured {
  import play.api.Play.current
  import akkaSystem.dispatcher

  def index = withAuth { username => implicit request =>
    Async {
      for {
        posts <- Post.all
        categories <- Category.titles
      }
      yield Ok(views.html.admin.Posts.index(posts.docs[Post])(categories, flash))
    }
  }

  def create = TODO

  def newPost = withAuth { username => implicit request =>
    Async {
      for(categories <- Category.titles)
      yield Ok(views.html.admin.Posts.edit(postForm)(categories))
    }
  }

  def edit(id:String) = withAuth { username => implicit request =>
    Async {
      for {
        post <- Post.getById(id)
        categories <- Category.titles
      }
      yield post.map { post =>
        Ok(views.html.admin.Posts.edit(postForm.fill(post))(categories))
      }.getOrElse(NotFound)
    }
  }

  def update(id:String) = TODO

  def delete(id: String) = withAuth { username => implicit request =>
    Async {
      for {
        categories <- Category.titles
        posts <- Post.all
        (_, post) <- Post.delete(id)
      }
      yield {
        Redirect(routes.Posts.index).flashing("success" -> s"""Post "${post.data.title}" successfully deleted.""")
      }
    }
  }

  import play.api.data.Form
  import play.api.data.Forms._

  val postForm = Form[Post](
    mapping(
      "title" -> nonEmptyText,
      "slug" -> nonEmptyText,
      "publishedAt" -> optional(date("yyyy-MM-dd")),
      "body" -> optional(text),
      "categories" -> list(text)
    )
      ((title, slug, publishedAt, body, categories) => Post(None, title, slug, publishedAt, body, Some(categories)))
      ((post: Post) => Some((post.title, post.slug, post.publishedAt, post.body, post.categories.getOrElse(Nil).toList)))
  )
}